src/main/java/com/coreos/jetcd/LeaseImpl.java

```diff
@@ -62,158 +65,192 @@
     }
+
+    this.keepAliveFuture.cancel(true);
+    this.deadlineFuture.cancel(true);
+    this.keepAliveRequestStreamObserver.onCompleted();
+    this.keepAliveResponseStreamObserver.onCompleted();
+    this.scheduledExecutorService.shutdownNow();
+    this.closeKeepAlives();
+  }
+
+  private synchronized void removeKeepAlive(long leaseId) {
+    this.keepAlives.remove(leaseId);
+  }
+
+  private void closeKeepAlives() {
+    final LeaseKeepAliveResponseWithError errResp = new LeaseKeepAliveResponseWithError(
+        new IllegalStateException("Lease client has closed"));
+    this.keepAlives.values().forEach(ka -> {
+      ka.sentKeepAliveResp(errResp);
+      ka.close();
+    });
+    this.keepAlives.clear();
+  }
+
+  private void start() {
+    this.sendKeepAliveExecutor();
+    this.deadLineExecutor();
+  }
+
+  private void reset() {
+    this.keepAliveFuture.cancel(true);
+    this.keepAliveRequestStreamObserver.onCompleted();
+    this.keepAliveResponseStreamObserver.onCompleted();
+    this.sendKeepAliveExecutor();
+  }
+
+  private void sendKeepAliveExecutor() {
+    this.keepAliveResponseStreamObserver = this.createResponseStreamObserver();
+    StreamObserver<LeaseKeepAliveRequest> requestStreamObserver = this.leaseStub
+        .leaseKeepAlive(this.keepAliveResponseStreamObserver);
+    this.keepAliveRequestStreamObserver = requestStreamObserver;
+    this.keepAliveFuture = scheduledExecutorService
+        .scheduleAtFixedRate(() -> {
+          long now = System.currentTimeMillis();
+
+          // send keep alive req to the leases whose next keep alive is before now.
+          this.keepAlives.entrySet().stream()
+              .filter(entry -> entry.getValue().getNextKeepAlive() < now)
+              .map((Entry::getKey))
+              .collect(Collectors.toList())
+              .forEach(leaseID -> requestStreamObserver
+                  .onNext(this.newKeepAliveRequest(leaseID)));
+        }, 0, 500, TimeUnit.MILLISECONDS);
+  }
+
+  private StreamObserver<LeaseKeepAliveResponse> createResponseStreamObserver() {
+    return new StreamObserver<LeaseKeepAliveResponse>() {
+      @Override
+      public void onNext(LeaseKeepAliveResponse leaseKeepAliveResponse) {
+        processKeepAliveResponse(leaseKeepAliveResponse);
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        processOnError();
+      }
+
+      @Override
+      public void onCompleted() {
+      }
+    };
+  }
+
+  private synchronized void processOnError() {
+    if (this.closed) {
+      return;
+    }
+
+    this.scheduledExecutorService.schedule(() -> reset(), 500, TimeUnit.MILLISECONDS);
+  }
+
+  private synchronized void processKeepAliveResponse(
+      LeaseKeepAliveResponse leaseKeepAliveResponse) {
+    if (this.closed) {
+      return;
+    }
+
+    long leaseID = leaseKeepAliveResponse.getID();
+    long ttl = leaseKeepAliveResponse.getTTL();
+
+    KeepAlive ka = this.keepAlives.get(leaseID);
+    if (ka == null) {
```

Comment:
When ka is closed due all the listeners are closed just before a ka response for the lease id arrives. In that case, don't broadcast ka response if the ka is deleted.